File: CPP/125.ValidPalindrome.cpp
#include <bits/stdc++.h>
using namespace std;

bool isPalindrome(string s) {
    string ns="";
    for (char ch:s) {
        if (isalnum(ch)) {
            ns+=tolower(ch);
        }
    }
    //cout << ns << endl;
    string rev=string(ns.rbegin(),ns.rend());
    return ns==rev;
}

int main() {
    cout << isPalindrome("A man, a plan, a canal: Panama") << endl; //true
    cout << isPalindrome("race a car")  << endl; //false
    cout << isPalindrome(" ") << endl; //true
    cout << isPalindrome("0P") << endl; //false
    return 0;
}

File: Python/125.ValidPalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        newstr=''
        for ch in s:
            if ch.isalnum():
                newstr+=ch.upper()

        #print(newstr)
        l,r=0,len(newstr)-1
        while l<r:
            if newstr[l]!=newstr[r]: return False
            l+=1
            r-=1
        return True
        

s=Solution()
print(s.isPalindrome("A man, a plan, a canal: Panama")) #true
print(s.isPalindrome("race a car")) #false
print(s.isPalindrome(" ")) # true